ingestion/src/metadata/utils/collaborative_super.py

```diff
@@ -0,0 +1,28 @@
+"""
+This module provides utilties for building a collaborative constructor library.
+"""
+
+from abc import ABC
+
+
+class Root:
+    """Root class for any class that needs to implement a colllaborative constructor but
+    might end up at the end of the inheritance chain. Since python's object.__init__ is not
+    a collaborative constructor, we need to have a root class that has a collaborative constructor.
+    """
+
+    __terminal__ = {object, ABC}
+
+    def __init__(self, *args, **kwargs):
```

Comment:
what's a collaborative constructor? "chaining" the inits?

might need an explanation on the code 🤔

Reply:
I added [tests to showcase](https://github.com/open-metadata/OpenMetadata/pull/18433/files#diff-8fd3265276fcfd4518c10ae6370c1665a6942c248da144996f81e6d460df435f) why we need this. I can also add a docstring but will be helpful to get some prompts as to what to document.

Reply:
There is a docstring in the delivery. Can you point to specific points that require clarification so I can expand it?

Reply:
might not be needed to change in the docstring, I was just trying to follow the logic of the implementation and I am getting a bit lost there

Reply:
which part?

Reply:
When I remove this class from these classes:
- SQAInterfaceMixin
- StoredStatisticsSource
- ProfilerInterface

I get this error when I try to run the postgres profiler:

```python
[2024-11-07 13:58:02] DEBUG    {metadata.Ingestion:status:92} - Traceback (most recent call last):
  File "/Users/imri/projects/OpenMetadata-pycharm/ingestion/src/metadata/ingestion/api/step.py", line 109, in run
    result: Either = self._run(record)
                     ^^^^^^^^^^^^^^^^^
  File "/Users/imri/projects/OpenMetadata-pycharm/ingestion/src/metadata/profiler/processor/processor.py", line 58, in _run
    profiler_runner: Profiler = record.profiler_source.get_profiler_runner(
                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/imri/projects/OpenMetadata-pycharm/ingestion/src/metadata/profiler/source/base/profiler_source.py", line 262, in get_profiler_runner
    profiler_interface = self.create_profiler_interface(
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/imri/projects/OpenMetadata-pycharm/ingestion/src/metadata/profiler/source/base/profiler_source.py", line 195, in create_profiler_interface
    profiler_interface: ProfilerInterface = profiler_class.create(
                                            ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/imri/projects/OpenMetadata-pycharm/ingestion/src/metadata/profiler/interface/profiler_interface.py", line 230, in create
    return cls(
           ^^^^
  File "/Users/imri/projects/OpenMetadata-pycharm/ingestion/src/metadata/profiler/interface/sqlalchemy/profiler_interface.py", line 97, in __init__
    super().__init__(
  File "/Users/imri/projects/OpenMetadata-pycharm/ingestion/src/metadata/profiler/interface/profiler_interface.py", line 140, in __init__
    super().__init__(
TypeError: object.__init__() takes exactly one argument (the instance to initialize)
```

This is the same `TypeError` that is demonstrated in the unit test. Any suggestions?